lib/Sema/DerivedConformances.cpp

```diff
@@ -380,3 +389,215 @@ bool DerivedConformance::checkAndDiagnoseDisallowedContext(
+                                  C.getIdentifier(indexName),
+                                  funcDecl);
+  indexVar->setInterfaceType(intType);
+  indexVar->setImplicit();
+
+  // generate: var indexVar
+  Pattern *indexPat = new (C) NamedPattern(indexVar, /*implicit*/ true);
+  indexPat->setType(intType);
+  indexPat = TypedPattern::createImplicit(C, indexPat, intType);
+  indexPat->setType(intType);
+  auto *indexBind = PatternBindingDecl::createImplicit(
+      C, StaticSpellingKind::None, indexPat, /*InitExpr*/ nullptr, funcDecl);
+
+  unsigned index = 0;
+  SmallVector<ASTNode, 4> cases;
+  for (auto elt : enumDecl->getAllElements()) {
+    // generate: case .<Case>:
+    auto pat = new (C) EnumElementPattern(TypeLoc::withoutLoc(enumType),
+                                          SourceLoc(), DeclNameLoc(),
+                                          DeclNameRef(), elt, nullptr);
+    pat->setImplicit();
+    pat->setType(enumType);
+
+    auto labelItem = CaseLabelItem(pat);
+
+    // generate: indexVar = <index>
+    auto indexExpr = buildIntegerLiteral(C, index++);
+
+    auto indexRef = new (C) DeclRefExpr(indexVar, DeclNameLoc(),
+                                        /*implicit*/true,
+                                        AccessSemantics::Ordinary,
+                                        LValueType::get(intType));
+    auto assignExpr = new (C) AssignExpr(indexRef, SourceLoc(),
+                                         indexExpr, /*implicit*/ true);
+    assignExpr->setType(TupleType::getEmpty(C));
+    auto body = BraceStmt::create(C, SourceLoc(), ASTNode(assignExpr),
+                                  SourceLoc());
+    cases.push_back(CaseStmt::create(C, SourceLoc(), labelItem, SourceLoc(),
+                                     SourceLoc(), body,
+                                     /*case body vardecls*/ None));
+  }
+
+  // generate: switch enumVar { }
+  auto enumRef = new (C) DeclRefExpr(enumVarDecl, DeclNameLoc(),
+                                     /*implicit*/true,
+                                     AccessSemantics::Ordinary,
+                                     enumVarDecl->getType());
+  auto switchStmt = SwitchStmt::create(LabeledStmtInfo(), SourceLoc(), enumRef,
+                                       SourceLoc(), cases, SourceLoc(), C);
+
+  stmts.push_back(indexBind);
+  stmts.push_back(switchStmt);
+
+  return new (C) DeclRefExpr(indexVar, DeclNameLoc(), /*implicit*/ true,
+                             AccessSemantics::Ordinary, intType);
+}
+
+/// Returns the ParamDecl for each associated value of the given enum whose type
+/// does not conform to a protocol
+/// \p theEnum The enum whose elements and associated values should be checked.
+/// \p protocol The protocol being requested.
+/// \return The ParamDecl of each associated value whose type does not conform.
+SmallVector<ParamDecl *, 3>
+swift::associatedValuesNotConformingToProtocol(DeclContext *DC, EnumDecl *theEnum,
```

Comment:
Can we limit the scope of this API promotion to DerivedConformance?

Reply:
done, also moved `convertEnumToIndex`, `enumElementPayloadSubpattern`, and `indexedVarDecl` into the same scope